### PR TITLE
Login hooks

### PR DESCRIPTION
**Pull request by ChristophWurst**
I found an inconsistency of login hooks while working on https://github.com/owncloud/core/pull/25154.
1. hooks were triggered twice (once in `login` and then in `loginWithToken`)
2. the post login hook in `login` was called without tranlating a possible token password to the login password
3. when logging in a disabled user via token no `LoginException` was thrown (the login was still blocked, but resulted in a 401 instead)
4. when logging in via token there was no check whether an app aborted the login process via the post login hook

~~Needs rebase once https://github.com/owncloud/core/pull/25154 was merged.~~ done

### CODE REVIEW

**Comment by ChristophWurst** [comment]
Note for reviewers: ``login`` handles the process of logging in by password, ``loginByToken`` handles the token login process. Hence they should perform the same checks.

**Comment by ChristophWurst** [comment]
- [x] TEST: LoginException is thrown in case a user was disabled
- [x] TEST: login with password works
- [x] TEST: login with token works

**Comment by guruz** [comment]
Code looks OK, I have not tested it.
👍

**Comment by ChristophWurst** [comment]
this fixes token login for clients, which seems to have issues on current master (see failing integration tests of other PRs). @DeepDiver1975 @PVince81 @icewind1991 review please

File: lib/private/User/Session.php [hunk 1]
```diff
@@ -280,46 +280,11 @@ public function getLoginName() {
...
-					$this->manager->emit('\OC\User', 'preLogin', array($uid, $loginPassword));
-				} catch (PasswordlessTokenException $ex) {
-					$this->manager->emit('\OC\User', 'preLogin', array($uid, ''));
-				}
-			} catch (InvalidTokenException $ex) {
-				// Invalid token, nothing to do
-			}
 
-			$this->loginWithToken($password);
-			$user = $this->getUser();
```
Comment (ChristophWurst) [inline lib/private/User/Session.php]: the return statement is missing on master, hence the ``return false;`` below is executed and the login is considered to have failed.

File: lib/private/User/Session.php [hunk 2]
```diff
@@ -449,6 +414,34 @@ public function tryBasicAuthLogin(IRequest $request) {
 		return false;
 	}
 
+	private function loginWithPassword($uid, $password) {
```
Comment (PVince81) [inline lib/private/User/Session.php]: PHPDoc maybe, including the thrown exception ?

File: lib/private/User/Session.php [hunk 3]
```diff
@@ -472,11 +467,22 @@ private function loginWithToken($token) {
...
+
+		if ($this->isLoggedIn()) {
+			$this->prepareUserLogin();
+		} else {
+			// injecting l10n does not work - there is a circular dependency between session and \OCP\L10N\IFactory
+			$message = \OC::$server->getL10N('lib')->t('Login canceled by app');
+			throw new LoginException($message);
+		}
 
 		$this->manager->emit('\OC\User', 'postLogin', array($user, $password));
```
Comment (PVince81) [inline lib/private/User/Session.php]: This is inconsistent: in the other method the postLogin hook is triggered **before** `if ($this->isLoggedIn())` but here it's done afterwards.

File: lib/private/User/Session.php [hunk 4]
```diff
@@ -449,6 +414,34 @@ public function tryBasicAuthLogin(IRequest $request) {
...
+		$user = $this->manager->checkPassword($uid, $password);
+		if ($user === false) {
+			// Password check failed
+			return false;
+		}
+
+		if ($user->isEnabled()) {
+			$this->setUser($user);
+			$this->setLoginName($uid);
+			$this->manager->emit('\OC\User', 'postLogin', array($user, $password));
```
Comment (PVince81) [inline lib/private/User/Session.php]: here the hook is triggered before `if ($this->isLoggedIn())`

**Comment by ChristophWurst** [comment]
PHPDoc added and ``postLogin`` hook moved accordingly.

**Comment by PVince81** [comment]
:-1: looks like token auth is broken. Works fine on master.

**Comment by PVince81** [comment]
Wait, I forgot to pull

**Comment by PVince81** [comment]
Tested with ext storage that uses post login for user provided creds (had to rebase onto master for that), both password and token login work and the ext storage can be accessed in both cases

:+1:

**Pull request merged**